### PR TITLE
Feat/application.housekeeping!

### PR DESCRIPTION
**Pull request by mejroslav**
**Implement publishing 'Application.housekeeping!'**

- By default this event triggers nightly at 03:00.
- A user can set the housekeeping time in config file:
```
[general]
housekeeping_time=HH:MM
```
- The app checks every 10 minutes if it is time for housekeeping. If so, it publishes 'Application.housekeeping!' and sets the next time for tomorrow.

### CODE REVIEW

File: asab/application.py [hunk 1]
```diff
@@ -584,3 +589,32 @@ def time(self):
...
+			seconds=now.second,
+			microseconds=now.microsecond)  # today at 00:00
+		next_housekeeping_time = td_midnight + datetime.timedelta(
+			hours=config_house_time.hour,
+			minutes=config_house_time.minute)  # today at the time for housekeeping
+		if now > next_housekeeping_time:
+			next_housekeeping_time += datetime.timedelta(days=1)
+		return next_housekeeping_time
+
+	def check_for_housekeeping(self, message_type):
```
Comment (ateska) [inline asab/application.py:614]: `_on_housekeeping_tick(...)`

File: asab/application.py [hunk 2]
```diff
@@ -584,3 +589,32 @@ def time(self):
...
+		"""Check if it's time for 'Application.housekeeping!'.
+		If so, publish the message and set housekeeping time for the next day.
+		"""
+		if datetime.datetime.now(datetime.timezone.utc) > self.NextHousekeeping:
+			self.PubSub.publish("Application.housekeeping!")
```
Comment (ateska) [inline asab/application.py:619]: - [ ] Comment this please
- [ ] Also ensure that the housekeeping is not executed immediately after app is started (after 3:00). The app has to wait till a first 3:00 ... or another check must be done that the current time is >3:00 AND <3:20 (for example).

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/application.py (hunk 2 above)
Comment (mejroslav) [reply asab/application.py:619]: The second one should be already checked in `set_housekeeping_time_from_config()`, see the lines 609-610.
```python
if now > next_housekeeping_time:
	next_housekeeping_time += datetime.timedelta(days=1)
```

File: asab/application.py [hunk 3]
```diff
@@ -584,3 +589,33 @@ def time(self):
...
+		if now > next_housekeeping_time:
+			next_housekeeping_time += datetime.timedelta(days=1)
+		return next_housekeeping_time
+
+	def _on_housekeeping_tick(self, message_type):
+		"""Check if it's time for 'Application.housekeeping!'.
+		If so, publish the message and set housekeeping time for the next day.
+		"""
+		if datetime.datetime.now(datetime.timezone.utc) > self.NextHousekeeping:
+			self.PubSub.publish("Application.housekeeping!")
```
Comment (byewokko) [inline asab/application.py:620]: I'd also add a LOG_NOTICE when housekeeping is triggered.

**Review by byewokko — CHANGES_REQUESTED** [review]

**Review by byewokko — APPROVED** [review]

File: asab/application.py [hunk 4]
```diff
@@ -584,3 +589,36 @@ def time(self):
 		Return UTC unix timestamp using a loop time (a fast way how to get a wall clock time).
 		'''
 		return self.BaseTime + self.Loop.time()
+
+
+	# Housekeeping
+
+	def set_housekeeping_time_from_config(self):
```
Comment (ateska) [inline asab/application.py:596]: `_set_housekeeping_time_from_config()` ... we don't want people to call this method

**Review by ateska — APPROVED** [review]

**Comment by ateska** [comment]
Extra points for documentation ;-)

**Pull request merged**